src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -13107,10 +13111,23 @@ private ParameterSyntax ParseLambdaParameter()
             var identifier = this.ParseIdentifierToken();
             ParseParameterNullCheck(out var exclamationExclamationToken, out var equalsToken);
 
-            // 'equalsToken' should always be null here. Otherwise, it means while we were scanning
-            // for potential lambdas, we saw an `!!=` or `!=` in the parameter list and allowed
-            // it. However, that should never happen as ScanParenthesizedLambda allows `!!` only.
-            Debug.Assert(equalsToken == null);
+            // If we didn't already consume an equals sign as part of !!=, then try to scan one out now. Note: this is
+            // not legal code.  But we detect it so that we can give the user a good message, and so we don't go
+            // completely off the rails.
+            //
+            // Note: we add the `= value` as skipped trivia to either the identifier or `!!` (if we have the latter).
+            // This allows us to handle this code without ever showing it the binding phases.  We could consider
+            // actually binding this in the future if it would be helpful and if we're ok paying the testing cost of
+            // checking this at the semantic layers.
```

Comment:
Consider dropping the end of the comment - we typically bind code the user wrote, even in error situations, so the cost shouldn't be an issue.
```suggestion
            // This allows us to handle this code without ever showing it to the binding phases.  We could consider
            // actually binding this in the future.
```

Reply:
> so the cost shouldn't be an issue.

In this case, i mentioned that it's "testing cost" :)  In other words, i'd like to have the better error recovery here, but i don't really want to write the tons of semantic tests that validate this :)